clients/bazel-module-registry/src/main/kotlin/BazelModuleRegistryService.kt

```diff
@@ -0,0 +1,169 @@
+
+import java.net.URI
+import java.util.Base64
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.Serializer
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonNamingStrategy
+
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.OkHttpClient
+
+import retrofit2.Retrofit
+import retrofit2.http.GET
+import retrofit2.http.Path
+
+import org.ossreviewtoolkit.model.Hash
+import org.ossreviewtoolkit.model.HashAlgorithm
+import org.ossreviewtoolkit.model.RemoteArtifact
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.model.VcsType
+
+@Serializer(URI::class)
+object URISerializer : KSerializer<URI> {
+    override fun serialize(encoder: Encoder, value: URI) = encoder.encodeString(value.toString())
+    override fun deserialize(decoder: Decoder) = URI(decoder.decodeString())
+}
+
+/**
+ * Interface for a Bazel Module Registry, based on the directory structure of https://bcr.bazel.build
+ * and the git repository it is based on (https://github.com/bazelbuild/bazel-central-registry/).
+ */
+interface BazelModuleRegistryService {
+    companion object {
+        /**
+         * The JSON (de-)serialization object used by this service.
+         */
+        private val JSON = Json {
+            ignoreUnknownKeys = true
+            namingStrategy = JsonNamingStrategy.SnakeCase
+        }
+
+        /**
+         * The service uses the Bazel Central Registry by default.
+         */
+        const val DEFAULT_URL = "https://bcr.bazel.build"
+
+        /**
+         * Create a Bazel Module Registry service instance for communicating with a server running at the given [url],
+         * defaulting to the Bazel Central Registry, optionally with a pre-built OkHttp [client].
+         */
+        fun create(
+            url: String = DEFAULT_URL,
+            client: OkHttpClient? = null
+        ): BazelModuleRegistryService {
+            val bmrClient = client ?: OkHttpClient()
+
+            val contentType = "application/json".toMediaType()
+            val retrofit = Retrofit.Builder()
+                .client(bmrClient)
+                .baseUrl(url)
+                .addConverterFactory(JSON.asConverterFactory(contentType))
+                .build()
+
+            return retrofit.create(BazelModuleRegistryService::class.java)
+        }
+    }
+
+    @Serializable
+    data class ModuleMetadata(
+        @Serializable(URISerializer::class) val homepage: URI?,
+        val maintainers: List<ModuleMaintainer>? = null,
+        val repository: List<String>? = null,
+        val versions: List<String>,
+        // The key in the map is the version, the value the reason for yanking it.
+        val yankedVersions: Map<String, String>,
+    ) {
+        fun vcsInfo(): VcsInfo {
```

Comment:
I have doubts this is a good location for this function. It maps the upstream model to ORT's model,
which I believe this should be place near by the code of the bazel analyzer.

Reply:
I moved them to the Analyzer plugin.